naga/src/back/msl/writer.rs

```diff
@@ -3720,6 +3730,734 @@ impl<W: Write> Writer<W> {
+                            (name, 12)
+                        }
+                        Sint32x4 => {
+                            let name = self.namer.call("unpackSint32x4");
+                            writeln!(
+                                self.out,
+                                "metal::int4 {name}(uint b0, \
+                                                    uint b1, \
+                                                    uint b2, \
+                                                    uint b3, \
+                                                    uint b4, \
+                                                    uint b5, \
+                                                    uint b6, \
+                                                    uint b7, \
+                                                    uint b8, \
+                                                    uint b9, \
+                                                    uint b10, \
+                                                    uint b11, \
+                                                    uint b12, \
+                                                    uint b13, \
+                                                    uint b14, \
+                                                    uint b15) {{"
+                            )?;
+                            writeln!(
+                                self.out,
+                                "{}return metal::int4(as_type<metal::int>(b0 << 24 | b1 << 16 | b2 << 8 | b3), \
+                                                      as_type<metal::int>(b4 << 24 | b5 << 16 | b6 << 8 | b7), \
+                                                      as_type<metal::int>(b8 << 24 | b9 << 16 | b10 << 8 | b11), \
+                                                      as_type<metal::int>(b12 << 24 | b13 << 16 | b14 << 8 | b15);",
+                                back::INDENT
+                            )?;
+                            writeln!(self.out, "}}")?;
+                            (name, 16)
+                        }
+                        Unorm10_10_10_2 => {
+                            let name = self.namer.call("unpackUnorm10_10_10_2");
+                            writeln!(
+                                self.out,
+                                "metal::float4 {name}(uint b0, \
+                                                      uint b1, \
+                                                      uint b2, \
+                                                      uint b3) {{"
+                            )?;
+                            writeln!(
+                                self.out,
+                                "{}return unpack_unorm10a2_to_float(b0 << 24 | b1 << 16 | b2 << 8 | b3);",
+                                back::INDENT
+                            )?;
+                            writeln!(self.out, "}}")?;
+                            (name, 4)
+                        }
+                    };
```

Comment:
All functions seem to do big-endian unpacking but the data in the buffers is meant to be unpacked as little-endian. This is covered in https://gpuweb.github.io/gpuweb/wgsl/#internal-value-layout (not very easy to find).

Reply:
Fixed.